### PR TITLE
Added e2e test for phone verification

### PR DESCRIPTION
**Pull request by sbryzak**
This PR adds a test for a successful phone verification workflow.

Fixes https://issues.redhat.com/browse/CRT-792

### CODE REVIEW

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [inline test/e2e/registration_service_test.go:497]: We are waiting for the default timeout here. So, this test will be always too slow.
Let's maybe do the following instead:
- Wait for StatusVerificationRequired: https://github.com/codeready-toolchain/host-operator/blob/9dc20a3c6c4522f916dd7db045421e7d3bc638bf/pkg/controller/usersignup/usersignup_controller.go#L281
- Check (without wait) that no MUR created/exist

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [inline test/e2e/registration_service_test.go:499]: Before that, let's also call the reg service /signup endpoint returns the expected status.

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [inline test/e2e/registration_service_test.go:512]: What body do we expect? Is it some JSON or it should be empty? Let's be specific and verify the body content if we expect any.

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [inline test/e2e/registration_service_test.go:550]: Let's also verify that the initial verification code from line 520 is still the same in UserSignup. Let's keep the code from line 520 and check that the same code in the update usersignup from line 541.
It would test that we do not update the code if verification failed.

**Review by alexeykazakov — COMMENTED** [review]
This test looks good overall (I have a few suggestions though).
But what about other use cases?
- When we exceed the max number of attempts.
- When the phone number already in use (including banned users).
- Invalid phone number.
- ...

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:493]: you know what is weird to me - that we firstly approve and then wait for the verification. Shouldn't that be vice versa? Approve only when it's verified?

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [reply test/e2e/registration_service_test.go:497]: For checking that it's not created you could use:
```
        err := s.hostAwait.WaitUntilMasterUserRecordDeleted(identity0.Username)
        require.NoError(t, err)
```

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:434]: The test has 170 lines... This is a pain to follow/understand - I got to the middle of the test and lost the whole context. Could you please break it down to either multiple subtests or multiple helper functions?

File: testsupport/user_setup.go
Comment (MatousJobanek) [inline testsupport/user_setup.go:111]: Just out of curiosity - what is the purpose of this change?

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [reply test/e2e/registration_service_test.go:493]: Yeah I guess that would be less ambiguous...

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [reply test/e2e/registration_service_test.go:497]: `WaitUntilMasterUserRecordDeleted()` is good for waiting for the MUR to be deleted:
T1. MUR exists.
T2. Do something which should delete the MUR.
T3. Wait for the MUR to not exist.

But it won't work properly for making sure that the MUR has never created:
T1. MUR does not exist.
T2. Do something which should prevent the MUR to be created
T3. Wait for the MUR to not exist (can pass even if the MUR is created with a delay).

File: testsupport/user_setup.go
Comment (sbryzak) [reply testsupport/user_setup.go:111]: We can configure certain email domains to not require phone verification - for testing purposes, we've configured both `redhat.com` and `acme.com` as excluded domains (see https://github.com/codeready-toolchain/host-operator/pull/291/files#diff-f9db715595ded24a97b8ca746f684769L7).  This change sets the email domain to one of the configured domains in order for the existing tests to still pass, as they would otherwise fail due to requiring verification.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:493]: That workflow actually seems a little weird to me, if I'm a user and I'm confronted with a phone verification requirement which I go to the effort to complete by providing my personal phone number, I can imagine being more than a little annoyed if after that I'm *still* required to wait for someone to manually approve me.  If I know that I need to wait for approval in the first place, then the phone verification step won't be a blocker for me to start using the product once someone has approved me.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:434]: It was a monumental effort to actually get the entire phone verification workflow tested and working successfully so yes, the size of the test did blow out a bit - I totally agree though that it's a bit too large to follow, so I'll make some changes to turn some of the duplicated code into reusable functions.

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [reply test/e2e/registration_service_test.go:493]: Phone verification is basically a prove that the user is not a bot. In many systems you can't even login without going through some verification. Including 2FA.
So, asking for phone verification (so the user is not a bot) before we even consider the user to be granted access to the system is IMO justified. We can double check with Parag though.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:434]: I've refactored it down to 100 lines, let me know if that looks ok now.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go]: Done.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go]: This has been refactored, now where a body is expected in the response the contents are checked.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go]: That would basically require all of this code yet again, when all we have done is manually approve the UserSignup.  We already have existing tests that check the status after a manual approval (see `TestSignupOK()`), do we really need to test this behaviour again here?

```	// Call get signup endpoint with a valid token and make sure it's pending approval
	body := invokeEndpoint(s.T(), "GET", s.route + "/api/v1/signup", token0, "", http.StatusOK)

	// Convert the signup response into a map so that we can examine its values
	mp := make(map[string]interface{})
	err = json.Unmarshal([]byte(body), &mp)
	require.NoError(s.T(), err)

	// Check that the response looks fine
	mpStatus, ok := mp["status"].(map[string]interface{})
	assert.True(s.T(), ok)
	assert.Equal(s.T(), "", mp["compliantUsername"])
	assert.Equal(s.T(), identity0.Username, mp["username"])
	require.IsType(s.T(), false, mpStatus["ready"])
	assert.False(s.T(), mpStatus["ready"].(bool))
	assert.Equal(s.T(), "PendingApproval", mpStatus["reason"])
	require.True(s.T(), mpStatus["verificationRequired"].(bool))

	// Now approve the usersignup.
	userSignup.Spec.Approved = true
	err = s.hostAwait.Client.Update(context.TODO(), userSignup)
	require.NoError(s.T(), err)```

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go]: I've already updated this code to what Matous suggested - is it the consensus that this won't actually work?

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go]: I would definitely like to double check on this, I still feel uncomfortable about forcing a user to provide their phone number with the possibility that they still might potentially need to wait for an extended period of time to be approved.  What about situations where we have run out of cluster resources - what is the expected process here and what will the delay be for the user?  What if there are thousands of users in the queue waiting to be approved?

Comment (sbryzak) [reply test/e2e/registration_service_test.go]: Also in response to the comment about many existing systems requiring verification, I am struggling to think of a single example in which there might still be a required manual approval step after already verifying.

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [reply test/e2e/registration_service_test.go]: No, I think we are good. I see that we do check `require.True(s.T(), mpStatus["verificationRequired"].(bool))` in the snippet above. So we are good.

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [reply test/e2e/registration_service_test.go]: `s.hostAwait.WaitUntilMasterUserRecordDeleted(identity0.Username) is unfortunately useless in this case. Take a look at my  previous message. So I think we have to implement it as a suggested in https://github.com/codeready-toolchain/toolchain-e2e/pull/167#discussion_r495219691.

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [reply test/e2e/registration_service_test.go]: We have manual approval step now for CRT cluster. I guess that we won't have it as soon as we go live. But it can happen that the clusters are full, then the user would need to wait.

Anyway, if I compare two types of the user experience:

First one that requires inputs from the user in two phases:
1. Sign up for the Dev Sandbox (that may require also creating a new user account at redhat.com)
2. wait until my account is approved (can be minutes, hours or days)
3. add additional information for phone verification (with the fact that the phone can be rejected since it was already used for some other user account)
4. Finally, enter Dev Sandbox

or provide all information at the beginning and then just wait until it is approved:
1. Sign up for the Dev Sendbox (that may require also creating a new user account at redhat.com) & verify the phone
2. Wait until my account is approved (can be minutes, hours or days)
3. Enter Dev Sandbox

To be honest, the second one seems to be more logical and fluent. I as a user give all necessary information about myself (including phone verification) at the beginning, then I just wait until I'm approved.

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [reply test/e2e/registration_service_test.go]: Yeah, I thought that checking that the MUR does not exist should be enough - for this particular use-case the method is sufficient - it verifies that he MUR does not exist. I know that ideally we should wait some short time and check that MUR is not really created, but since we are going to wait for StatusVerificationRequired (as suggested by Alexey), then we should be fine.

File: testsupport/user_setup.go
Comment (MatousJobanek) [reply testsupport/user_setup.go:127]: Ok, I missed the connection. Thanks a lot for the explanation.

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [reply test/e2e/registration_service_test.go]: There is a couple of cases where we don't have a choice and have to wait for some resource to be not be created. For example in https://github.com/codeready-toolchain/toolchain-e2e/blob/4996af7b68436dfe74338ef560c8e0cb60804060/test/e2e/usersignup_test.go#L133
But we introduced `WithRetryOptions()` to override the default timeout (which is too long) for such cases:
```
s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second * 10)).WaitForMasterUserRecord(username)
require.Error(s.T(), err)
```
It still slows down the test but not for the entire default timeout (10 sec vs 60 sec per call). But in this particular case I think we could just wait for StatusVerificationRequired and don't wait for MUR at all. So we can save those 10 seconds.

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [reply test/e2e/registration_service_test.go]: +1 @MatousJobanek

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go]: I've removed the MUR check here and replaced it with a UserSignup status check, plus a check the confirm that `verificationRequired` is correctly set to true when calling the getSignup endpoint

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go]: Actually looking at the code, this conversation is moot.... if verification is required then the `verificationRequired` property will be set to true regardless of whether the usersignup is approved or not, meaning that a new user should immediately have to perform phone verification after signing up anyway.  I've modified the test workflow to do the approval after the verification has been completed, in order to reflect this.

File: test/e2e/registration_service_test.go
Comment (alexeykazakov) [reply test/e2e/registration_service_test.go]: But we still need to check that the MUR is not created. Why not checking that it does not exist?

Comment (sbryzak) [reply test/e2e/registration_service_test.go]: Your very last comment was "But in this particular case I think we could just wait for StatusVerificationRequired and don't wait for MUR at all." What did you mean by this?

Comment (alexeykazakov) [reply test/e2e/registration_service_test.go]: If we wait for StatusVerificationRequired then we don't have to **wait** for the MUR to. In other words we don't have to use any `WaitFor...()` function and can check if the MUR exists or not without any timeout/retries/waits. Sorry for not being clear.

Comment (alexeykazakov) [reply test/e2e/registration_service_test.go]: Just try to load the MUR from the client. It should not found any.

Comment (sbryzak) [reply test/e2e/registration_service_test.go]: No problem, your use of the word "just" in that context made me think you only cared about the status check.  I'll add a lookup then for the MUR and expect an error there.

File: test/e2e/registration_service_test.go
Comment (tinakurian) [inline test/e2e/registration_service_test.go:464]: /@sbryzak it looks like there isn't a test for verifying that the user can request one too many new verification codes? this should probably be added

Comment (tinakurian) [reply test/e2e/registration_service_test.go:464]: @sbryzak it would be nice if we could better organize the tests as well - it's a bit hard to follow

File: test/e2e/registration_service_test.go
Comment (tinakurian) [inline test/e2e/registration_service_test.go:476]: @sbryzak shouldn't we also test that when the expiry happens we get the correct result?

File: test/e2e/registration_service_test.go
Comment (tinakurian) [inline test/e2e/registration_service_test.go:465]: @sbryzak after looking at the tests i guess we don't test any invalid phone number entries - maybe we should add that?

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:470]: Doesn't this test in registration service already do this?

https://github.com/codeready-toolchain/registration-service/blob/master/pkg/controller/signup_test.go#L365

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:482]: I thought we had this covered already in registration service, if not we should probably add a test for it there.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:471]: Same as above, this test doesn't require user interaction beyond registration service so doesn't need to be an end to end test, if we don't test this in registration service already then we should probably add it there.

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [reply test/e2e/registration_service_test.go]: Yes, you are right. :+1: 
I'm going to swap the verification & approval phases in my PR anyway: https://github.com/codeready-toolchain/host-operator/pull/288/files#diff-5b4b18a0fc2e0d708e4e6334cdb6bc69L285-L294 so I'm glad that we agreed on the flow

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:444]: I have maybe a dumb question - couldn't we use the `postSignup` function for that - it looks like it's doing almost the same https://github.com/codeready-toolchain/toolchain-e2e/blob/master/testsupport/user_setup.go#L124-L148
We could technically reuse the `invokeEndpoint` function there.

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:467]: It's just a detail, but we should check that the error is `NotFound` - something like this:
```suggestion
	require.Error(s.T(), err)
	require.True(s.T(), errors.IsNotFound(err))
```

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:475]: If I understand it correctly, we send `PUT` request to the registration service with the country code and phone number and the reg-service then updates the UserSignup. I know that it may look like a detail, but it could happen that we get the UserSignup in the test earlier than registration service updates it - I was just fixing such flaky tests https://github.com/codeready-toolchain/toolchain-e2e/pull/172 so I'd like to avoid it for the future. 
WDYT? Could you create a `UserSignupWaitCriterion` for that?

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:488]: The same case is also here.

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:492]: btw, do you think that we could verify the expected number (and also in other cases of the annotations)? or at least verify that the number has increased since the last time...?

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:502]: same case with the wait also here

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [inline test/e2e/registration_service_test.go:528]: if we want to check that it's provisioned then we should verify the existence of the condition:
```suggestion
	_, err = s.hostAwait.WaitForMasterUserRecord(identity0.Username, wait.UntilMasterUserRecordHasCondition(Provisioned()))
```

**Review by MatousJobanek — COMMENTED** [review]
Thanks a lot for changing the structure of the tests - this is much better.
I have a few questions/proposals, but nothing significant. Good job :+1:

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:444]: I could use it, but it's not going to reduce the lines of code - the email claim and the token are still required as they are used further down in the test.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:467]: No problem, added.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:475]: I'm not quite sure what you mean here.  The init verification (i.e. the PUT request) is synchronous, it will not return a response until after the UserSignup has been updated.  The `invokeEndpoint()` function is also synchronous, it will not return until it has fully read the body of the response.  How is it possible that we could get the `UserSignup` in the test earlier than the registration service updates it?

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:492]: We could add this test to registration service, I'll make a note to do it when I take care of the architectural refactor.  I discussed this very issue (of trying to cover every possible scenario in the e2e tests) with Alexey today and he agreed with me that the e2e tests was not the right place to do this.  If there's certain conditions that we don't currently test for in registration service then we should add a new test there, not in the e2e tests.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:528]: Good idea, updated.

File: test/e2e/registration_service_test.go
Comment (sbryzak) [reply test/e2e/registration_service_test.go:494]: For reference - I've added a reminder to add a test to cover this here: https://issues.redhat.com/browse/CRT-825

File: testsupport/conditions.go
Comment (sbryzak) [reply testsupport/conditions.go]: Good catch, I've added it.

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [reply test/e2e/registration_service_test.go:477]: ah, right - I totally forgot that it is synchronous (I'm still living in the asynchronous world :grimacing:) Thanks for explanation :+1:

File: test/e2e/registration_service_test.go
Comment (MatousJobanek) [reply test/e2e/registration_service_test.go:494]: cool, thanks :+1:

**Review by MatousJobanek — APPROVED** [review]
Thanks for addressing my comments :+1:

File: test/e2e/registration_service_test.go
Comment (tinakurian) [reply test/e2e/registration_service_test.go:472]: yep it does, but the tests in registration-service are heavily mocked - it would be nice to have some that aren't mocked imo

**Pull request merged**
